### PR TITLE
Delete all the associated provisioned users when IDP deletes

### PR DESCRIPTION
**Pull request by GANGANI**
### Description

Delete all the associated provisioned users when IDP deletes

### Approach

Assign thread to perform associated provisioned users delete task asynchronously from the IDP delete operation.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            FrameworkUtils.startTenantFlow(tenantDomain);
+            RealmService realmService = FrameworkServiceComponent.getRealmService();
+            int tenantId = realmService.getTenantManager().getTenantId(tenantDomain);
+            UserRealm userRealm = realmService.getTenantUserRealm(tenantId);
+            String[] userList =
+                    ((AbstractUserStoreManager)userRealm.getUserStoreManager()).
+                            getUserList(FrameworkConstants.PROVISIONED_SOURCE_CLAIM, resourceId, null);
+            for (String username : userList) {
+                ((AbstractUserStoreManager) userRealm.getUserStoreManager()).deleteUser(username);
+            }
```
Comment (somindatommy) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java:75]: Let's add a log after completing the operation successfully.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    private void provisionedUserDelete() {
+
+        try {
+            FrameworkUtils.startTenantFlow(tenantDomain);
+            RealmService realmService = FrameworkServiceComponent.getRealmService();
+            int tenantId = realmService.getTenantManager().getTenantId(tenantDomain);
+            UserRealm userRealm = realmService.getTenantUserRealm(tenantId);
+            String[] userList =
+                    ((AbstractUserStoreManager)userRealm.getUserStoreManager()).
+                            getUserList(FrameworkConstants.PROVISIONED_SOURCE_CLAIM, resourceId, null);
```
Comment (somindatommy) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java:72]: Shall we check for null objects?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            for (String username : userList) {
```
Comment (somindatommy) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java:73]: Shall we stop before the `for loop` if we returned a null or empty list

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/listener/JITProvisioningIdentityProviderMgtListener.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    public boolean doPostDeleteIdPByResourceId(String resourceId, IdentityProvider identityProvider,
+                                               String tenantDomain) throws IdentityProviderManagementException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("doPostDeleteIdPByResourceId executed for idp: " + resourceId + " of tenant domain: " +
+                    tenantDomain);
+        }
+
+        if (!FrameworkUtils.isEnhancedFeature()) {
+            // JIT provisioning enhanced feature is not enabled. Hence returning.
```
Comment (somindatommy) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/listener/JITProvisioningIdentityProviderMgtListener.java:53]: Shall we add a debug log here? WDYT?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 5]
```diff
@@ -3035,4 +3035,17 @@ public static void resetAuthenticationContext(AuthenticationContext context) {
...
+     * Check whether the JIT provisioning enhanced feature is enabled.
+     *
+     * @return true if the JIT provisioning enhanced features is enabled else return false.
+     */
+    public static boolean isEnhancedFeature() {
+
+        if (StringUtils.isNotBlank(IdentityUtil.getProperty(FrameworkConstants.ENABLE_ENHANCE_FEATURE))) {
+            return Boolean.parseBoolean(IdentityUtil.getProperty(FrameworkConstants.ENABLE_ENHANCE_FEATURE));
+        }
+        return false;
```
Comment (somindatommy) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:3049]: Can't we just use `Boolean.parseBoolean`? won't it return `false` when the value if null?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java (hunk 1 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java:75]: Added here[1]

[1] https://github.com/wso2/carbon-identity-framework/pull/3597/files#diff-21403306f5143e0d531bbb02e8f8167a17337dfc4ec20bb67228791b6d68024aR57-R60

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/listener/JITProvisioningIdentityProviderMgtListener.java (hunk 4 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/listener/JITProvisioningIdentityProviderMgtListener.java:53]: We maintain this config throughout the implementation of this JIT provisioning enhanced feature. IMO logging this information will not be required here. WDYT?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/listener/JITProvisioningIdentityProviderMgtListener.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    public boolean doPostDeleteIdP(String idPName, String tenantDomain) throws IdentityProviderManagementException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("doPostDeleteIdp executed for idp: " + idPName + " of tenant domain: " + tenantDomain);
+        }
+
+        if (!FrameworkUtils.isEnhancedFeature()) {
+            // JIT provisioning enhanced feature is not enabled. Hence returning.
+            return true;
+        }
```
Comment (mevan-karu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/listener/JITProvisioningIdentityProviderMgtListener.java:87]: Execute `enhancedFeatureEnabled` check before the debug log or change the log message appropriately.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java [hunk 7]
```diff
@@ -38,6 +38,7 @@
     public static final String ACCOUNT_DISABLED_CLAIM_URI = "http://wso2.org/claims/identity/accountDisabled";
     public static final String ACCOUNT_UNLOCK_TIME_CLAIM = "http://wso2.org/claims/identity/unlockTime";
     public static final String USERNAME_CLAIM = "http://wso2.org/claims/username";
+    public static final String PROVISIONED_SOURCE_CLAIM = "http://wso2.org/claims/identity/userSourceId";
```
Comment (mevan-karu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java:41]:
```suggestion
    public static final String PROVISIONED_SOURCE_ID_CLAIM = "http://wso2.org/claims/identity/userSourceId";
```
WDYT?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java [hunk 8]
```diff
@@ -104,6 +105,9 @@
     public static final String IDP_ID = "idpId";
     public static final String ASSOCIATED_ID = "associatedID";
 
+    public static final String IDP_RESOURCE_ID = "IDPResourceID";
```
Comment (mevan-karu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java:108]:
```suggestion
    public static final String IDP_RESOURCE_ID = "idpResourceID";
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 9]
```diff
@@ -3035,4 +3035,17 @@ public static void resetAuthenticationContext(AuthenticationContext context) {
         context.setRetrying(false);
         context.setCurrentAuthenticator(null);
     }
+
+    /**
+     * Check whether the JIT provisioning enhanced feature is enabled.
+     *
+     * @return true if the JIT provisioning enhanced features is enabled else return false.
+     */
+    public static boolean isEnhancedFeature() {
```
Comment (mevan-karu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:3044]: Better to say what enhanced feature we're enabling in the method name.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java [hunk 10]
```diff
@@ -104,6 +105,9 @@
...
+    public static final String ENABLE_ENHANCE_FEATURE = "JITProvisioning.EnableEnhancedFeature";
```
Comment (mevan-karu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java:109]:
```suggestion
    public static final String ENABLE_JIT_PROVISIONING_ENHANCE_FEATURE = "JITProvisioning.EnableEnhancedFeature";
```
WDYT?

Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3597/commits/6e1589981f43782e7824b22d065010c4ef34b01d

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java (hunk 9 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3597/commits/6e1589981f43782e7824b22d065010c4ef34b01d

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java (hunk 8 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3597/commits/6e1589981f43782e7824b22d065010c4ef34b01d

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java (hunk 7 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3597/commits/6e1589981f43782e7824b22d065010c4ef34b01d

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/listener/JITProvisioningIdentityProviderMgtListener.java (hunk 6 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/listener/JITProvisioningIdentityProviderMgtListener.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3597/commits/6e1589981f43782e7824b22d065010c4ef34b01d

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java (hunk 5 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3597/commits/6e1589981f43782e7824b22d065010c4ef34b01d

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java (hunk 3 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3597/commits/6e1589981f43782e7824b22d065010c4ef34b01d

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java (hunk 2 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3597/commits/6e1589981f43782e7824b22d065010c4ef34b01d

**Review by mevan-karu — APPROVED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        this.tenantDomain = tenantDomain;
+    }
+
+    @Override
+    public void run() {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Start running the JIT provisioned user delete task.");
+        }
+        diagnosticLog.info("Start running the JIT provisioned user delete task.");
```
Comment (IsuraD) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java:56]: lets remove diagnosticLog logs here.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    private void provisionedUserDelete() {
+
+        try {
+            FrameworkUtils.startTenantFlow(tenantDomain);
+            RealmService realmService = FrameworkServiceComponent.getRealmService();
+            int tenantId = realmService.getTenantManager().getTenantId(tenantDomain);
+            UserRealm userRealm = realmService.getTenantUserRealm(tenantId);
+            String[] userList =
```
Comment (IsuraD) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java:71]: There is a max limit of returning the users. What happens if it exceeds?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/listener/JITProvisioningIdentityProviderMgtListener.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+    public boolean doPreDeleteIdP(String idPName, String tenantDomain) throws IdentityProviderManagementException {
+
+        if (!FrameworkUtils.isJITProvisionedEnhancedFeatureEnabled()) {
+            // JIT provisioning enhanced feature is not enabled. Hence returning.
+            return true;
+        }
+
+        String idpId = IdentityProviderManager.getInstance().getIdPByName(idPName, tenantDomain,
+                true).getResourceId();
+        IdentityUtil.threadLocalProperties.get().put(FrameworkConstants.IDP_RESOURCE_ID, idpId);
```
Comment (IsuraD) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/listener/JITProvisioningIdentityProviderMgtListener.java:74]: What happens if there is an issue in the idp deletition. Then the doPostDeleteIdP will not call and the thread-local may not clear

Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/listener/JITProvisioningIdentityProviderMgtListener.java:75]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/3597/commits/0bb10a0d01ccec6f3ef0ba7ed7675e4fe360d177. Cleared the thread-local before retrieving it.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java (hunk 12 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/3597/commits/e520700cb02b5c9b67249b8e60c5cf5951dbe109.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java (hunk 11 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/3597/commits/0bb10a0d01ccec6f3ef0ba7ed7675e4fe360d177

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/JITProvisionedUserDeleteThread.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Thread to perform JIT provisioned users deletion task based on provisioned IDP delete.
+ */
+public class JITProvisionedUserDeleteThread implements Runnable {
+
+    private static final Log log = LogFactory.getLog(SessionCleanUpService.class);
```
Comment (ashensw) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/JITProvisionedUserDeleteThread.java:45]:
```suggestion
    private static final Log log = LogFactory.getLog(JITProvisionedUserDeleteThread.class);
```

Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/JITProvisionedUserDeleteThread.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3597/commits/624b4876b35b8ed350ebddafb26607512c2cc5e8

**Review by ashensw — APPROVED** [review]

**Pull request merged**